go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -114,44 +102,47 @@ func (npw *nodePortWatcher) AddService(service *kapi.Service) {
 						npw.gwBridge, stderr, err, flow)
 				}
 			}
-		}
-		for _, externalIP := range service.Spec.ExternalIPs {
-			if err := util.ValidatePort(svcPort.Protocol, svcPort.Port); err != nil {
-				klog.Errorf("Skipping service add for svc: %s, err: %v", svcPort.Name, err)
-				continue
-			}
-			flowProtocol := protocol
-			nw_dst := "nw_dst"
-			if utilnet.IsIPv6String(externalIP) {
-				flowProtocol = protocol6
-				nw_dst = "ipv6_dst"
-			}
-			_, stderr, err := util.RunOVSOfctl("add-flow", npw.gwBridge,
-				fmt.Sprintf("priority=100, in_port=%s, %s, %s=%s, tp_dst=%d, actions=%s",
-					npw.ofportPhys, flowProtocol, nw_dst, externalIP, svcPort.Port, npw.ofportPatch))
-			if err != nil {
-				klog.Errorf("Failed to add openflow flow on %s for ExternalIP: "+
-					"%s, stderr: %q, error: %v", npw.gwBridge, externalIP, stderr, err)
+
+			for _, externalIP := range service.Spec.ExternalIPs {
+				if err := util.ValidatePort(svcPort.Protocol, svcPort.Port); err != nil {
+					klog.Errorf("Skipping service add for svc: %s, err: %v", svcPort.Name, err)
+					continue
+				}
+				flowProtocol := protocol
+				nwDst := "nw_dst"
+				if utilnet.IsIPv6String(externalIP) {
+					flowProtocol = protocol6
+					nwDst = "ipv6_dst"
+				}
+				_, stderr, err := util.RunOVSOfctl("add-flow", npw.gwBridge,
+					fmt.Sprintf("priority=100, in_port=%s, %s, %s=%s, tp_dst=%d, actions=%s",
+						npw.ofportPhys, flowProtocol, nwDst, externalIP, svcPort.Port, npw.ofportPatch))
```

Comment:
this looks like a good candidate for an OVS API, no? ovs.AddFlows(port, proto, dst, src, ...)
out of the scope of the PR but we can add a TODO , it' s a "good first issue" for people jumping into the project

Reply:
yeah we have a JIRA open downstream for someone to refactor the openflow manager and make it better. @rcarrillocruz is assigned to it. Created an upstream issue to track there too: https://github.com/ovn-org/ovn-kubernetes/issues/1888